### ISSUE TITLE
deserialize with type parameter

### ISSUE THREAD

**Issue by tpapp** [JuliaLang/julia#20066]
(I asked [on discourse](https://discourse.julialang.org/t/cant-deserialize-serialized-data/1500), didn't get a reply there, didn't find anything that looks like a duplicate, so I am opening an issue).

On `v0.5.0`, `Base.deserialize` fails with the following error message:
```julia
ERROR: TypeError: Array: in parameter, expected Type{T}, got Dict{Int64,Dict{Int64,
Array{T,1}}}                                                                      
 in deserialize_datatype(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}
}) at ./serialize.jl:830                                                          
 in handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}},
 ::Int32) at ./serialize.jl:571                                                   
 in deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at ./s
erialize.jl:541                                                                   
 in collect_to!(::Array{DataType,1}, ::Base.Generator{UnitRange{Int64},Base.Seriali
zer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}, ::Int64, ::I
nt64) at ./array.jl:340                                                           
 in collect(::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationSt
ate{Base.AbstractIOBuffer{Array{UInt8,1}}}}}) at ./array.jl:308                   
 in deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Type
{SimpleVector}) at ./serialize.jl:588                                             
 in handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}},
 ::Int32) at ./serialize.jl:581                                                   
 in deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at ./s
erialize.jl:541                                                                   
 in deserialize_datatype(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}
}) at ./serialize.jl:829                                                          
 in handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}},
 ::Int32) at ./serialize.jl:571                                                   
 in deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at ./s
erialize.jl:541                                                                   
 in deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Type
{Dict{Int64,Dict{Int64,Array{T,1}}}}) at ./serialize.jl:894                       
 in deserialize_datatype(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}
}) at ./serialize.jl:835                                                          
 in handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}},
 ::Int32) at ./serialize.jl:571                                                   
 in deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at ./s
erialize.jl:541                                                                   
 in roundtrip(::Dict{Int64,Dict{Int64,Array{T,1}}}) at ./REPL[374]:5
```
using the following minimal working example:
```julia
function generate(T)
    dd = Dict{Int,Dict{Int,T}}()
    for i in 1:10
        dd[i] = Dict{Int,T}()
        for j in 1:10
            dd[i][j] = rand(Int,10)
        end
    end
    dd
end

function roundtrip(data)
    io = IOBuffer(true, true)
    serialize(io, data)
    seekstart(io)
    deserialize(io)
end

roundtrip(generate(Vector{Int})) # OK
roundtrip(generate(Vector))      # fails, error message above
```

**Comment by alexmorley** [comment]
Because Vector is a TypeConstructor while Vector{Int} is a DataType? Vector{Any} will work...

**Comment by vtjnash** [comment]
on master, I get:
```julia
julia> roundtrip(generate(Vector))
ERROR: KeyError: key 1 not found
Stacktrace:
 [1] getindex; at associative.jl:262 [inlined]
 [2] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:570
 [3] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:544
 [4] collect(::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}) at array.jl:381
 [5] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Type{SimpleVector}) at serialize.jl:591
 [6] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:584
 [7] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:544
 [8] deserialize_datatype(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:840
 [9] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:574
 [10] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::DataType) at serialize.jl:899
 [11] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:584
 [12] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:544
 [13] collect_to!(::Array{DataType,1}, ::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}, ::Int64, ::Int64) at array.jl:414
 [14] collect_to_with_first!(::Array{DataType,1}, ::Type{T} where T, ::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}, ::Int64) at array.jl:401
 [15] collect(::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}) at array.jl:382
 [16] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Type{SimpleVector}) at serialize.jl:591
 [17] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:584
 [18] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:544
 [19] deserialize_datatype(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:840
 [20] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:574
 [21] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:544
 [22] collect_to!(::Array{DataType,1}, ::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}, ::Int64, ::Int64) at array.jl:414
 [23] collect_to_with_first!(::Array{DataType,1}, ::Type{T} where T, ::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}, ::Int64) at array.jl:401
 [24] collect(::Base.Generator{UnitRange{Int64},Base.Serializer.##3#4{SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}}}) at array.jl:382
 [25] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Type{SimpleVector}) at serialize.jl:591
 [26] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:584
 [27] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:544
 [28] deserialize_datatype(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:840
 [29] handle_deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}, ::Int32) at serialize.jl:574
 [30] deserialize(::SerializationState{Base.AbstractIOBuffer{Array{UInt8,1}}}) at serialize.jl:544
 [31] roundtrip(::Dict{Int64,Dict{Int64,Array{T,1} where T}}) at REPL[2]:5
 [32] eval(::Module, ::Any) at ./boot.jl:236
 [33] eval(::Module, ::Any) at /Users/jameson/julia/usr/lib/julia/sys.dylib:?
 [34] eval_user_input(::Any, ::Base.REPL.REPLBackend) at REPL.jl:66
 [35] macro expansion; at REPL.jl:97 [inlined]
 [36] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at event.jl:71
```

**Comment by JeffBezanson** [comment]
Seems to be due to an incomplete fix for #15163.
